### ISSUE TITLE
[STORM-992] A bug in the timer.clj might cause unexpected delay to schedule new event

### ISSUE THREAD

**Issue by jira-importer** [apache/storm#4830]
<p>The timer thread calculates the delay to schedule the head of the queue and sleeps accordingly. However, if a new event with high priority is inserted at the head of the queue during the sleep, this event cannot be scheduled in time.</p>

---
<details><summary><i>Originally reported by <a title="Li Wang" href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=wangli1426">wangli1426</a>, imported from: <a href="https://issues.apache.org/jira/browse/STORM-992" target="_blank">A bug in the timer.clj might cause unexpected delay to schedule new event</a></i></summary>
<i><ul>
<li><b>assignee</b>: <a title="Li Wang" href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=wangli1426">wangli1426</a>
<li><b>status</b>: Resolved
<li><b>priority</b>: Major
<li><b>resolution</b>: Fixed
<li><b>resolved</b>: 2015-08-25T01:27:13+01:00
<li><b>imported</b>: 2025-01-24
</ul></i>
</details>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=wangli1426">wangli1426</a>:</i>
<p>This bug has been fixed at <a href="https://github.com/apache/storm/pull/680" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/680</a></p>

<p>Please consider to merging it.</p>

<p>Thanks.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user wangli1426 commented on the pull request:</p>

<p>    <a href="https://github.com/apache/storm/pull/680#issuecomment-130932282" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/680#issuecomment-130932282</a></p>

<p>    @knusbaum I have filed a jira at <a href="https://issues.apache.org/jira/browse/STORM-992" class="external-link" rel="nofollow">https://issues.apache.org/jira/browse/STORM-992</a>.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user knusbaum commented on the pull request:</p>

<p>    <a href="https://github.com/apache/storm/pull/680#issuecomment-130933080" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/680#issuecomment-130933080</a></p>

<p>    +1 </p>

<p>    I should've mentioned it, but we prefix our pull request titles with the jira id:<br/>
    ```<br/>
    <a href="https://issues.apache.org/jira/browse/STORM-992" title="A bug in the timer.clj might cause unexpected delay to schedule new event" class="issue-link" data-issue-key="STORM-992"><del>STORM-992</del></a> A bug in the timer.clj might cause unexpected delay to schedule new event<br/>
    ```</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user wangli1426 commented on the pull request:</p>

<p>    <a href="https://github.com/apache/storm/pull/680#issuecomment-130939524" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/680#issuecomment-130939524</a></p>

<p>    @knusbaum I have changed the title of this pull request as suggested. I am sorry for your inconvenience. </p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=kabhwan">kabhwan</a>:</i>
<p>It will be closed when PR is merged to Apache repo. Reopening.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=wangli1426">wangli1426</a>:</i>
<p>This bug has been fixed at <a href="https://github.com/apache/storm/pull/680" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/680</a>, waiting for being merged.<br/>
I believe this fix could prevent potential bugs from happening. So please merge it.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user asfgit closed the pull request at:</p>

<p>    <a href="https://github.com/apache/storm/pull/680" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/680</a></p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user HeartSaVioR commented on the pull request:</p>

<p>    <a href="https://github.com/apache/storm/pull/680#issuecomment-134423270" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/680#issuecomment-134423270</a></p>

<p>    +1.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=kabhwan">kabhwan</a>:</i>
<p>Since it incurs potential bug which would be hard to debug, I also merged this to 0.10.x-branch.</p>

<p>Thanks, <a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=wangli1426" class="user-hover" rel="wangli1426">Li Wang</a><br/>
I merged this to master and 0.10.x-branch.</p>
